Runs crawler service under a dedicated system account

conorsch opened the pull request:
Reorganizes the provisioning flow to accommodate deployments in multi-host and multi-user environments. Implements the improvements discussed in #51, viz. 
- project directory moved to `/opt/FingerprintSecureDrop`
- runs as `fpsd` system account, _except under vagrant_, where `vagrant` is still used to facilitate development
- writes general database config to global env, but database passwords to users' home directories

In the process, all Ansible provisioning was folded into a single role, so vars can be trivially overridden in different environments, and the method for updating vars is intuitive and DRY. The new provisioning flow preserves the current behavior in the Vagrant developer environment, but provisioning to a multi-host setup is now simpler, thanks to the judicious use of blocks to configure any combination of the following:
- crawler code
- database backend
- database client credentials

There's currently a TODO in the README about documenting the multi-host setup. Happy to take care of that, but I think it'd best be handled as a separate PR. 

To try out the new setup, destroy and recreate the VM under this setup. In particular, use both the `./crawler.py` and `sudo systemctl start crawler` invocations during debugging. I've seen both those methods exit early (less than a minute of run time), so clearly more work needs to be done. Flagging for you review before I do any more work.

conorsch committed:
Namespaces vars in common and crawler roles

Temporary, pending merging the two roles, but this makes things cleaner.

conorsch committed:
Configures and builds tor source as system account

Rather than having multiple developers maintaining multiple copies of
the tor source code, let's download, verify, configure and build tor as
a normal user, using the dedicated system account `fpsd`.

conorsch committed:
Runs crawler service as system account `fpsd`

In order to support a system-wide service configuration, the crawler
should run under a dedicated account. We'll default to `fpsd`, and
override at the play level for vagrant. We can't chown/chgrp when
running under vagrant, since the project directory will be mounted in.

Converts crawler systemd service file to template

Allows variable interpolation, e.g. project root, and soon to come
environment variables.

Uses system account to clone git repo

We want the /opt/FingerprintSecureDrop directory to be writable by the
system account that runs the `crawler` service, so let's set ownership
and permissions accordingly.

The system account must exist prior to the clone operation, so
reorganizing the task list accordingly.

conorsch committed:
Installs and runs TBB as system account for crawler service

The Tor Browser bundle is designed to be run as a normal user from
within a home directory. Intsalling the software system-wide is not
necessarily the best approach.

conorsch committed:
Consolidates package installation tasks in crawler role

conorsch committed:
Folds common and sorter roles into crawler role

The sorter role had a single task in it, installing Python dependencies
via pip from a requirements.txt file. The location of the
requirements.txt was determined by a previous role, which is dirty, so
let's merge the task lists so we can reuse vars.

The apt packages, pip dependencies, and git cloning tasks were
previously housed in the `common` role, and are now in the crawler role
where they belong.

conorsch committed:
Consolidates database setup task list

The database role was overloaded, and configured both server and client
connections. Let's make that distinction clearer by isolating the task
lists according to the intended use case. That way, we can reuse the
role and its vars for configuring clients, as well.

conorsch committed:
Moves project directory to /opt

By default the role will clone the git repository into /opt, but let's
provide a toggle var to skip cloning in when running via vagrant, since
the directory will already be mounted there.

Uses var for declaring paths to requirements files
Installing via pip in a loop now, reusing the var for the base
directory, to keep config changes DRY.

Updates logging filepath in tor source patch
Now using the /opt/FingerprintSecureDrop base directory.

conorsch committed:
Supports writing postgres credentials for multiple users

Rather than only configuring the PGPASS file for a single user, let's
default to the user running Ansible, but support a list of usernames to
configure access for. The role must support targeting individual
usernames, because postgres mandates that PGPASS files be mode 600.

conorsch committed:
Folds database role into into crawler role

Specifically, the following tasks are now ported over:

  * schema setup
  * pgpass templating
  * environment variable settings

The schema tasks now use a fileglob, so all the schema files shipped
with the role will automatically be applied.

Storing the database config tasks in the same role that deploys the
crawler code allows clear and concise reuse of variables.

conorsch committed:
Writes database env vars to crawler service config

Using a discrete file with restricted permissions so only root and the
crawler service account can read the database auth env vars.

conorsch committed:
More durable changed status on tor-src dependencies

The logic for the deb-src parameter in the apt module inappropriately
reports changed=yes always, so let's inspect the output and report
accordingly. Previously the inspection mandated that all packages were
fully up to date, but that's not an accurate representation of changed
status for this task: only if packages were installed or upgraded should
the task report changed.

conorsch committed:
Simplifies database password generation logic

Previously the role used multiple tasks and complicated Jinja template
logic to figure out what the database password should be. Let's condense
that conditional logic into a single var, `fpsd_database_password`, that
will be reused throughout the role. By default, that var will be
autogenerated by Ansible. As a result, the role will remain idempotent
as long as the /tmp/passwordfile exists on the Ansible
controller; thereafter, it'll be regenerated and the associated tasks
will correctly report "changed". Overriding the `fpsd_database_password`
var will result in full idempotence even across reboots.

conorsch committed:
Makes crawler service configuration optional

By default the role will still apply the crawler code tasks, but it's
now possible to skip them via a var in order to run just the database or
credential config tasks. This allows the same role and vars setup to be
distributing across multiple hosts when running the crawler in
production.

conorsch committed:
Boosts RAM for dev VM 512MB -> 1024MB

Compiling source code warrants a GB of RAM to keep things snappy.

conorsch commented:
@fowlslegs r? 

We should get these changes in prior to beginning work on #56, otherwise we're looking at rebasing until the cows come home.

conorsch commented:
@msheiny Take a look over the Ansible reorg and let me know if you have strong opinions. One of the primary goals is to make deployment to the prod machines sane, especially since we're using multiple servers there for crawler fleet, database, and ML tooling.

One strong opinion I have about a decent dev-to-prod workflow is accurate representation of the architecture when working locally. That would entail creating a multi-VM Vagrantfile, which isn't a challenge technical, but may be prohibitive in the sense that it would consume a couple GBs of RAM.

fowlslegs commented:
Sorry about merging #58 and making you rebase, but the separate requirements files were getting unwieldy and pip only functions right at 98.6 degrees Fahrenheit.

fowlslegs commented:
The playbook repeatedly failed for me. I tried every immediately conceivable way to upload the traceback to Github, but it just wouldn't go. So I uploaded it to https://linx.li/hu8iom6f.txt.

conorsch commented:
> The playbook repeatedly failed for me. I tried every immediately conceivable way to upload the traceback to Github, but it just wouldn't go.

Sounds like a networking problem on your host. The traceback certainly indicates networking issues, and if you weren't able connect reliably to GitHub, the Ansible config likely isn't the issue. "Works on my machine":

```
TASK [crawler : Install and upgrade Python packaging and debugging tools.] *****
task path: /home/conor/freedomofpress/FingerprintSecureDrop/roles/crawler/tasks/packages.yml:21
changed: [default] => (item=pip) =>

[snip]
```

Using Ansible 2:

```
ansible --version                                                       
ansible 2.1.0.0                                                                                                           
  config file =                                                                                                                                  
  configured module search path = Default w/o overrides
```

Give it another go after destroying the VM.

fowlslegs commented:
We're now specifying a particular Ansible (and pip-tools) version. See https://github.com/freedomofpress/FingerprintSecureDrop#getting-started.

That said it was a networking thing... buildling.

psivesely commented on roles/crawler/defaults/main.yml:
```diff
@@ -1,20 +1,91 @@
 
 # Specify Tor version to download, patch, build, and install
-tor_release: 0.2.8.7
+fpsd_crawler_tor_release: 0.2.8.7
```
Can you bump this to 0.2.8.8?

conorsch replied:
There's yet another pending rebase on #66 that should land first; say when and I'll push it up.

psivesely commented on roles/crawler/defaults/main.yml:
```diff
@@ -1,20 +1,91 @@
+fpsd_crawler_tbb_release: 6.0.5
+fpsd_crawler_tbb_locale: en-US
+fpsd_crawler_tbb_arch: 64
+fpsd_crawler_tbb_directory: "~{{ fpsd_crawler_system_account }}/tbb"
```
I think we should place this in `/opt/tbb`.

conorsch replied:
Doing so would have side-effects for the crawler code, given the tight coupling of the tbb path. TBB is also designed to be run out of a home directory with normal-user privileges, so this seems like the least rock-the-boat approach.

psivesely replied:
One would still be running it as a normal user. Either your own account, for manually running the Crawler, or as the `fpsd_crawler_system_account` (hopefully unprivileged), for running as a systemd unit.

A bit of a segue, but related--It'd be nice if the `fpsd_crawler_system_account` was created, even in Vagrant VMs, as a new user "crawler" by default. It's hard to evaluate what you've done here without knowing how you intend to combine it with the code you have (or have planned to write) in our infra repo. I'm guessing you intend to do something like that with play-levels tasks.

Anyway, IMO opinion it's nicer that all the code is under the `/opt` umbrella instead of any particular user's homedir. And it's not tightly coupled--`sed -i 's/expanduser("~")/"\/opt"/g' crawler.py` should do the trick.

conorsch replied:
> A bit of a segue, but related--It'd be nice if the `fpsd_crawler_system_account` was created, even in Vagrant VMs, as a new user "crawler" by default.

So the system account is currently called `fpsd`, and vagrant-specific logic just reuses the `vagrant` account as the service account. I can modify it so the `fpsd` account is used in all contexts.

psivesely replied:
I think it would be best to use the `fpsd` account in all contexts; it should only add a few seconds to provisioning Vagrant VMs, and will mirror the remote deployment more closely.

conorsch replied:
See my comment [here](https://github.com/freedomofpress/FingerprintSecureDrop/pull/54#issuecomment-253647295):

> The `vagrant` user is still used as the service account under local VMs, rather than `fpsd`. This is required, because the crawler code assumes it can write to its own directory, in order to append to local log files, and under vagrant, the code directory is owned by user `vagrant` and cannot be chowned due to its shared file system with the host OS.

So switching to the `fpsd` account under vagrant would mean losing live-mounting of the local directory, which would make local development a real chore.

psivesely commented on roles/crawler/defaults/main.yml:
```diff
@@ -1,20 +1,91 @@
 # Specify Tor version to download, patch, build, and install
-tor_release: 0.2.8.7
+fpsd_crawler_tor_release: 0.2.8.7
+fpsd_crawler_tor_directory: "~{{ fpsd_crawler_system_account }}"
```
Should also be in `/opt`. We don't want to have to build and/or download TBB for multiple users.

conorsch replied:
Writing to `/opt` would need to be done with elevated privileges, and actually `/usr/local/src` would be a more appropriate location here for a system-wide dir. Using the home dir, however, is designed to provide an unprivileged process for downloading, configuring, and compiling the source code; only the install command should involved privilege escalation.

> We don't want to have to build and/or download TBB for multiple users.

The approach of writing to the home dir does not require recompilation; even in multi-user setups, it will only build tor once.

psivesely replied:
Having recently read Debian and several other sources FHS guides, I would agree on `/usr/local/src` being the best place to build Tor. Though I don't get your point on needing elevated privileges for `/opt`--you'd just create a dir there that you'd set separate permissions for.

fowlslegs commented:
> I've seen both those methods exit early (less than a minute of run time), so clearly more work needs to be done. Flagging for you review before I do any more work.

https://github.com/freedomofpress/FingerprintSecureDrop/pull/59 was maybe part of that.

conorsch commented:
>#59 was maybe part of that.

Great! I'll rebase this on top of both #58 and #59, then.

fowlslegs commented:
The other problem being that our fixed EntryNode seems to be down.

conorsch commented:
Rebased on top of latest master to incorporate changes from both #58 and #59. I recommend running a fresh provision just to sanity-check the config; I'll be doing the same, and will report back with any problems you haven't already reported.

psivesely reviewed (changes requested):
Didn't heavily scrutinize this--assume it's mostly the same code moved around. 

Besides the comments, a couple more notes:
- As I mentioned in person the `configure-databases.yml` tasklisk isn't running even though `fpsd_database_initialize_local_db` is set to true by default. I tried the `fpsd_database_initialize_local_db | bool` syntax in that `when:` directive with no luck.
- From a security standpoint it would be best to re-write the playbook s.t. when you use it just to provision a remote database server it doesn't create a PGPASSFILE on that server (or run the `set-environment-variables.yml` tasklist at all). Instead just use the postgres account to do everything related to setting the database up. That will still work for the Vagrant VM setting too.

conorsch committed:
Bumps tor version 0.2.8.7 -> 0.2.8.8

Per @fowlslegs's review of #54.

conorsch committed:
Downloads tor source to /opt/tor

Rather than download and compile the tor source in the `fpsd` user's
home directory, we'll place it in /opt, with ownership set to `fpsd`.

conorsch committed:
Downloads tbb to /opt/tbb

Rather than download the Tor Browser Bundle to the `fpsd` user's
home directory, we'll place it in /opt, with ownership set to `fpsd`.

Requires changes to the hardcoded path within the `crawler.py` code, so
updated that as well to point to the new path.

conorsch committed:
Implements a restart delay for the crawler service

When restarting the service, wait 10 seconds before triggering the new
start, to allow for clean up of calls made to external programs such as
tbb.

conorsch commented:
Pushed up a few changes to respond to review. In sum:

* bumps tor version 0.2.8.7 -> 0.2.8.8
* tor source dir now targets `/opt/tor`
* tbb install dir now targets `/opt/tbb`
* added a 10-second grace period on crawler service restarts, to allow for cleanup

The `vagrant` user is still used as the service account under local VMs, rather than `fpsd`. This is required, because the crawler code assumes it can write to its own directory, in order to append to local log files, and under vagrant, the code directory is owned by user `vagrant` and cannot be chowned due to its shared file system with the host OS.

No changes were made to the logging setup for the crawler code. While it would be nice to lean on systemd's handling of logging functionality, doing so would require refactoring how the crawler handles logging (basically, tell it to spit logs to stdout), and that doesn't seem worth the effort.

Had trouble duplicating the database tasks being skipped—those tasks run fine on my system. Try again with the new changes here and let me know if the problem persists for you, @fowlslegs.

fowlslegs commented:
> No changes were made to the logging setup for the crawler code. While it would be nice to lean on systemd's handling of logging functionality, doing so would require refactoring how the crawler handles logging (basically, tell it to spit logs to stdout), and that doesn't seem worth the effort.

It's actually pretty easy with a `logging.StreamHandler` to do this. I'll probably use this and `argparse`, so that the service file can pass a `--service` flag that switches logging from logfiles to stdout. Merging this should come first though.

> The vagrant user is still used as the service account under local VMs, rather than fpsd. This is required, because the crawler code assumes it can write to its own directory, in order to append to local log files, and under vagrant, the code directory is owned by user vagrant and cannot be chowned due to its shared file system with the host OS.

The logging locations could be changed in the future (e.g., `/var/log/tor/cells.log`, if we use a stream-handler we can avoid the crawler writing to `fpsd/logging`, etc.) but definitely don't worry about it now. I do like the idea of mirroring the remote deployment as closely as possible, but not worth the effort for now.

> tor source dir now targets /opt/tor

In my followup to your comment I agreed `/usr/local/src` would be the better place.

> /usr/local/src : Source code (place where to extract and build non debian'ized stuffs)

-- https://wiki.debian.org/FilesystemHierarchyStandard

psivesely commented on roles/crawler/defaults/main.yml:
```diff
@@ -1,20 +1,91 @@
 # Specify Tor version to download, patch, build, and install
-tor_release: 0.2.8.7
+fpsd_crawler_tor_release: 0.2.8.8
+fpsd_crawler_tor_directory: "/opt/tor"
```
`/usr/local/src/tor`

conorsch replied:
Handled in 58bdcf.

psivesely commented on roles/crawler/tasks/build-install-tor.yml:
```diff
@@ -1,34 +1,28 @@
+  - name: Clobber relay.c in Tor {{ fpsd_crawler_tor_release }} with custom relay.c from repo
+    patch:
+      src: relay.c.patch
+      # Not using the ~username prefix to the fullpath because the patch module
```
You can get rid of this comment now.

conorsch replied:
Handled in c6ee6d3.

conorsch committed:
Moves tor target directory to /usr/local/src/tor

It's source code, so put it where source code belongs.

conorsch committed:
Removes inaccurate comment about tor target directory

The tor source directory was previously within a home directory, and the
comment referenced the home expansion syntax. That's moot now, so we're
downloading and building tor in /usr/local/src now.

fowlslegs commented:
You need to rename `roles/crawler/files/postgres-schemas` to `roles/crawler/files/postgres`. That's why the task that creates the `raw` schema tables is failing. Furthermore, you can't use a fileglob because the tables need to be created in a particular order.

conorsch committed:
Explicitly order the postgres schema includes

We don't want to use a fileglob, because we need to ensure a specific
order of the schemas being applied. Besides, the fileglob path was
specified incorrectly.

conorsch commented:
Updated database schema import to include files in an explicit order, and with the correct filepath.

fowlslegs commented:
All tests passing! Good to go!
